Modules/md5module.c

```diff
@@ -177,7 +187,18 @@ MD5Type_update(MD5object *self, PyObject *obj)
 
     GET_BUFFER_VIEW_OR_ERROUT(obj, &buf);
 
-    update(self->hash_state, buf.buf, buf.len);
+    if (self->lock == NULL && buf.len >= HASHLIB_GIL_MINSIZE) {
+        self->lock = PyThread_allocate_lock();
+    }
+    if (self->lock != NULL) {
+        Py_BEGIN_ALLOW_THREADS
+        PyThread_acquire_lock(self->lock, 1);
```

Comment:
why does this not do what ENTER_HASHLIB does? (try to acquire the lock in a non-blocking way, then try again, but blocking)

Reply:
This is the instance constructor, so there can be no other references to the class or parallel calls yet.  thus no need to bother initializing the lock.

Reply:
oh, sorry, commented on the wrong part of the code.  comment deleted.

Reply:
I added a better comment to hashlib.h: basically the ENTER and LEAVE macros are intended to surround "fast" operations where dropping the GIL isn't worthwhile - they're solely about acquiring self->lock if defined, not about dropping the GIL.  So they use the non-blocking acquire, drop-gil if not & block to acquire and re-obtain the GIL, pattern.

here we're calling update() with the GIL released under the assumption that it does meaningful computation.